libjcat/jcat-bt-util.c

```diff
@@ -144,3 +146,405 @@ jcat_bt_generate_key_pair(const gchar *keyname,
 	return TRUE;
 #endif
 }
+
+/**
+ * jcat_bt_parse_private_key:
+ * @private_key_file_content: (not nullable): the entire file content of the private key
+ * @parsed_private_key: (not nullable) (out): the buffer for the private key
+ * @parsed_public_key: (not nullable) (out): the buffer for the public key
+ * @parsed_key_name: (not nullable) (out): the name of the key
+ * @parsed_key_hash: (not nullable) (out): the 32-bit hash of the key
+ * @error: #GError, or %NULL
+ *
+ * Parse a private key in the same format as Go's sumdb. This works with the output of
+ *jcat_bt_generate_key_pair.
+ *
+ * Since: 0.1.12
+ **/
+gboolean
+jcat_bt_parse_private_key(GBytes *private_key_file_content,
+			  GByteArray **parsed_private_key,
+			  GByteArray **parsed_public_key,
+			  gchar **parsed_key_name,
+			  gchar **parsed_key_hash,
+			  GError **error)
+{
+#ifndef ENABLE_ED25519
+	g_set_error_literal(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_FAILED,
+			    __FUNCTION__ " requires Ed25519 support but it is not enabled");
+	return FALSE;
+#else
+
+	const guint8 *data = NULL;
+	gsize size = 0;
+	const guint8 *next = NULL, *key_hash = NULL;
+	gsize key_name_size = 0;
+	static const gsize key_hash_size = 8;
+	g_autoptr(GString) base64_private = NULL;
+	g_autofree guchar *private = NULL;
+	gsize private_len = 0;
+	g_autoptr(GByteArray) public_key_for_validation = NULL;
+	g_autofree gchar *expected_key_hash = NULL;
+	gsize i;
+
+	static const gchar prefix[] = "PRIVATE+KEY+";
+
+	/* Looks like this: "PRIVATE+KEY+test+3d18b207+AXLw43DfQRIa8AB0FwAyP0clTh437+DCXuAg4FUb55LI"
+	 */
+
+	g_return_val_if_fail(parsed_private_key != NULL, FALSE);
+	g_return_val_if_fail(parsed_public_key != NULL, FALSE);
+	g_return_val_if_fail(parsed_key_name != NULL, FALSE);
+	g_return_val_if_fail(parsed_key_hash != NULL, FALSE);
+
+	data = g_bytes_get_data(private_key_file_content, &size);
+	if (size < sizeof prefix) {
+		g_set_error_literal(error,
+				    G_IO_ERROR,
+				    G_IO_ERROR_FAILED,
+				    "too short to be a private key");
+		return FALSE;
+	}
+	if (memcmp(data, prefix, sizeof prefix - 1) != 0) {
+		g_set_error_literal(
+		    error,
+		    G_IO_ERROR,
+		    G_IO_ERROR_FAILED,
+		    "unexpected private key format; must start with \"PRIVATE+KEY+\"");
+		return FALSE;
+	}
+
+	data += sizeof prefix - 1;
+	size -= sizeof prefix - 1;
+
+	/* The next field is the key name. */
+	next = memchr(data, '+', size);
+	if (next == NULL) {
+		g_set_error_literal(error,
+				    G_IO_ERROR,
+				    G_IO_ERROR_FAILED,
+				    "cannot find end of key name");
+		return FALSE;
+	}
+	if (*parsed_key_name) {
+		g_free(*parsed_key_name);
+	}
```

Comment:
not this; this is `out` only -- i.e. don't expect reuse.

Reply:
Done.